text/0007-service-modules.md

```diff
@@ -0,0 +1,125 @@
+* Start Date: 2022-08-31
+* RFC Type: decision
+* RFC PR: https://github.com/getsentry/rfcs/pull/7
+
+# Summary
+
+This RFC proposes a file structure for 'compartmentalized services' or 'domain
+boundaries'. The concept of service boundaries was introduced in 0002, and this
+document aims to provide more detailed guidelines for how 'services' in the
+monolith would be structured as Python modules.
+
+# Motivation
+
+The sentry monolith continues to grow in scope as we build new product features.
+As the application has grown the number of models, endpoints and tasks makes
+understanding how the application is inter-connected more challenging. The
+current code layout complicates optimizing CI, and impairs our ability to
+clearly dilineate product boundaries within the monolith.
+
+This RFC does not attempt to define what the boundaries and services within the
+monolith should be. Nor does it attempt to describe the organization of
+Typescript code.
+
+# Background
+
+Currently the Sentry monolith is organized as a single Django application that
+follows a typical project layout organized by 'kind of class'. For example, all
+models are co-located in a small number of directories, as are all endpoints and
+serializers. While this repository layout has served us well, it is increasingly
+hard to navigate as the application grows. At time of writing, we have:
+
+* ~275 endpoint modules
+* 115 model modules
+* 105 serializer modules
+
+Knowing how each of these classes are related to features in sentry is not
+always obvious. A similar problem exists for tests as there is no way to easily
+locate all the tests that need to be run when a model class changes.
+
+# Proposed Python Structure
+
+As sentry is a django application, we can leverage the
+[Django-apps](https://docs.djangoproject.com/en/4.1/ref/applications/) to act as
+a container for application services in the future. While not all services will
+need all the features of Django Applications, many will.
+
+## Django app structure
+
+We'll use 'discover' as an example for the service modules
+
+```
+src/sentry/discover
+    __init__.py
+    app.py
+    urls.py
+    models/__init__.py
+    models/discoversavedquery.py
```

Comment:
I think we should try to prevent the code of a service to directly access models from another service and doing joins. Joining DB tables across modules would defy the idea of creating boundaries.
When referencing the module class this is not different than preventing a service from accessing the internals of another one, though with Django you can run raw SQL queries. Do we allow that to begin with, and if yes, should we prevent it ?

Reply:
> I think we should try to prevent the code of a service to directly access models and tables from another service by doing joins across service boundaries. Joining DB tables across modules would defy the idea of creating boundaries.

I agree. I considered recommending not exporting Django models from a service. The problem came up against though was that python's typehints will want an importable class to use in typehints, which would generally need to be the model :cry: